### PR TITLE
Add a class TaskResult.

### LINKED ISSUE #16198: A TaskResult class.
Here's something I've wanted for a long time already, and current discussions in #16182 suggest that this is the right time:

#16182 adds a `Lazy<T>` class that allows us to declare members that are not initialized by default, but only upon first use. This is useful for large tables, for example, that may not be used in all programs -- the use case that inspired #16182 is in the finite element classes where we may or may not need the restriction or prolongation matrices, and only want to compute them if a program actually requests them.

A separate, but perhaps related, case is that a class would like to compute some information on a background thread. An example is a lot of the work we do for #15375 in `AffineConstraints::close()`. It would be nice if `close()` could start a thread that does the work of this function in the background; whenever one accesses a variable that is computed by `close()` one would then have to ensure that the background thread has finished.

`Lazy<T>` does not help with this because one can not ask whether its values have finished computing. We'd need a separate class (say `BackgroundThreadResult`) that encodes the semantics that there is a thread somewhere that is initializing the variable (the initialization may or may not have started yet -- but it's queued on some thread or task at least), and asking for the value of the object requires waiting for that thread or task to finish. Something like this:
```
template <typename T>
class BackgroundThreadResult {
public:
  BackgroundThreadResult (Task<T> &task);

  T & get() {
    task.join();
    return task.return_value();
  }
};
```
This could be made more efficient by moving the result of the task into a `std::optional` object once the `join()` operation has finished for the first time, at which point we can do `task = {}` to release all memory associated with the task.

Alternative designs would let the background task emplace its result in the object itself, and the emplacement would wake up everyone who is currently waiting in `get()` for the result to arrive. This could be done relatively straightforwardly via a `std::condition_variable` object. The only problem with that is that we would not be communicating to the task scheduler which task we are waiting for (which `task.join()` does), and consequently (i) the task scheduler does not see that we are currently blocked and so cannot schedule in other tasks, (ii) it is conceivable that we deadlock if the task scheduler only runs a certain number of tasks at the same time.

@tamiko Thoughts, since you are also currently thinking about these kinds of things?

### PR DESCRIPTION
**Pull request by bangerth**
I've wanted to have this class for a long time already and have a couple of places in mind where I'm going to use it. The most difficult part was to get the semantic around assignment right.

Closes #16198.

### CODE REVIEW

**Comment by bangerth** [comment]
Ping?

File: include/deal.II/base/task_result.h [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+            ExcMessage(
+              "You cannot assign a new task to a TaskResult object "
+              "while it is still waiting for its previous task "
+              "to finish. See the documentation of this class' operator=() "
+              "for more information."));
+      }
+    // Having established that there is no previous task still running,
+    // clear previous content of this class and set the current task
+    // as the one we're waiting for:
+    clear();
```
Comment (tjhei) [inline include/deal.II/base/task_result.h:373]: The code in this function above this call is duplicated in clear() except for the error message, right? Maybe that would be cleaner to call clear() here.

File: include/deal.II/base/task_result.h [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+    task_result = std::move(other.task_result);
+    other.task_result.reset();
+  }
+
+
+
+  template <typename T>
+  inline TaskResult<T>::~TaskResult()
+  {
+    // First ensure that there is no currently running task. As
```
Comment (tjhei) [inline include/deal.II/base/task_result.h:325]: why don't you just call clear() here?

File: include/deal.II/base/task_result.h [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+  inline const T &
+  TaskResult<T>::value() const
+  {
+    if (result_is_available)
+      return task_result.value();
+    else
+      {
+        wait_and_move_result();
+        return task_result.value();
+      }
```
Comment (tjhei) [inline include/deal.II/base/task_result.h:455]:
```suggestion
    if (!result_is_available)
      wait_and_move_result();
    return task_result.value();
```

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
Do we need to add documentation to ``Task`` mention this new class?

**Comment by bangerth** [comment]
Thanks, @tjhei, for the review. I have simplified the code as you suggested, and also updated the documentation of both `Task` and `TaskResult`.

**Review by tjhei — APPROVED** [review]

**Pull request merged**